oak-core/src/test/java/org/apache/jackrabbit/oak/plugins/index/AsyncIndexUpdateTest.java

```diff
@@ -307,6 +307,142 @@ public void testAsyncPause() throws Exception {
+                    String afterTime, AsyncUpdateCallback callback, AtomicReference<String> checkpointToRelease) throws CommitFailedException {
+                if (before == MISSING_NODE) {
+                    reindexingCount.incrementAndGet();
+                }
+                return super.updateIndex(before, beforeCheckpoint, after, afterCheckpoint, afterTime, callback, checkpointToRelease);
+            }
+        };
+        async.setLeaseTimeOut(250);
+        async.run();
+
+        builder = store.getRoot().builder();
+        builder.child("test").setProperty("foo", "b");
+        builder.child("child").setProperty("prop", "value");
+        store.merge(builder, EmptyHook.INSTANCE, CommitInfo.EMPTY);
+
+        merge.drainPermits();
+        merge.release(2);
+
+        // the first async.run() was a reindexing, but we're not counting that, so:
+        reindexingCount.set(0);
+
+        // System.out.println("Going for an async index run, reindex count = " + reindexingCount.get());
+
+        // with only 2 permits for the 'merge' semaphore,
+        // ie 2 calls to merge(), the async.run() below
+        // will hit the disposed NodeStore and trigger the bug.
+        // however, it will have released the checkpoint, as that
+        // is unprotected. so a subsequent call to async.run()
+        // will fail
+        Thread t = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    async.run();
+                } catch(Throwable th) {
+                    // th.printStackTrace();
+                }
+            }
+        });
+        t.start();
+        t.join(5000);
+
+        // System.out.println("Async index run done - restarting NodeStore, reindex count = " + reindexingCount.get());
+
+        // lets resume the NodeStore
+        merge.release(1000);
+        isDisposed.set(false);
+
+        // and rerun the async
+        // System.out.println("and going for another async index run, reindex count = " + reindexingCount.get());
+        async.run();
+        // we did crash before, so the async lease would still be valid, so we expect a failed run:
+        assertTrue(async.isFailing());
+
+        // let's wait until the lease is released - should be 500ms (2x lease timeout of 250ms)
+        // System.out.println("Looping... ");
+        final long timeout = System.currentTimeMillis() + 1000;
+        while(async.isFailing() && System.currentTimeMillis() < timeout) {
+            // let's be a bit CPU friendly..:
+            Thread.sleep(50);
+            // System.out.println("  Rechecking async index..., reindex count = " + reindexingCount.get());
+            async.run();
+            // System.out.println("  Result of async indexing: failing=" + async.isFailing() + " , reindex count = " + reindexingCount.get());
+        }
+        // the test timeout was hit
+        assertTrue(async.isFailing());
```

Comment:
What could be added here is an assertion on the reindexCount still being 0 (as that should be the case as it failed), eg
```suggestion
        assertTrue(async.isFailing());

        // but now should not have seen a complete reindex
        assertEquals("reindex happened", 0, reindexingCount.get());
```